### PR TITLE
feat: Add cross-platform build docker images and use quay.io/opendevin as image registry and org.

### PR DESCRIPTION
**Pull request by iFurySt**
Xnip2024-03-25_00-44-33
Xnip2024-03-25_00-44-39

### CODE REVIEW

File: evaluation/SWE-bench/Makefile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+REGISTRY=quay.io
```
Comment (rbren) [inline evaluation/SWE-bench/Makefile:1]: Per discussion in slack, should we switch to GitHub's ghcr.io as the registry?

**Comment by iFurySt** [comment]
feat: Add workflow for building and pushing multi-arch images to ghcr.io.

I have successfully tested the workflow in my repo, see the results here:
https://github.com/iFurySt/OpenDevin/actions/runs/8415626483/job/23041038547

screenshot in package:
Xnip2024-03-25_13-54-28

I'm not in the OpenDevin org, so you need to confirm the workflow permission was already set to RW. 
Your Org -> OpenDevin -> Actions -> Workflow permissions:
Xnip2024-03-25_13-33-27

**Comment by xingyaoww** [comment]
@huybery can you help take a look? I did not see this option on my Action panel

**Comment by huybery** [comment]
@iFurySt @xingyaoww I just opened it. How about now?

**Comment by iFurySt** [comment]
@huybery just review pr and merge it, the workflow will auto-trigger to build and push to OpenDevin's package, and maybe you need to public the packages after the images are pushed.

**Comment by iFurySt** [comment]
@xingyaoww @huybery I've just noticed the workflow triggered when pr is opened, I removed the trigger of pull_request in case the wrong pr will trigger the workflow to push the wrong images to ghcr

File: evaluation/SWE-bench/scripts/run_docker_interactive.sh [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-DOCKER_IMAGE=opendevin/eval-swe-bench:v0.1
+DOCKER_IMAGE=quay.io/opendevin/eval-swe-bench:v0.1
```
Comment (xingyaoww) [inline evaluation/SWE-bench/scripts/run_docker_interactive.sh:3]: Should this be changed to `ghcr.io`?

File: opendevin/sandbox/sandbox.py [hunk 3]
```diff
@@ -11,7 +11,8 @@
 InputType = namedtuple("InputType", ["content"])
 OutputType = namedtuple("OutputType", ["content"])
 
-CONTAINER_IMAGE = os.getenv("SANDBOX_CONTAINER_IMAGE", "opendevin/sandbox:latest")
+CONTAINER_IMAGE = os.getenv("SANDBOX_CONTAINER_IMAGE", "quay.io/opendevin/sandbox:latest")
```
Comment (xingyaoww) [inline opendevin/sandbox/sandbox.py:14]: same here, should it be `ghcr.io`?

File: .github/workflows/ghcr.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    inputs:
+      reason:
+        description: 'Why manual trigger?'
+        required: false
+        default: ''
+
+jobs:
+  ghcr_build_and_push:
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push'
```
Comment (xingyaoww) [inline .github/workflows/ghcr.yml:16]: Does this mean the workflow will build an image and push it to `ghcr.io` on every pushed commit? Is it possible to do this only when the `Dockerfile` is changed?

File: opendevin/sandbox/sandbox.py (hunk 3 above)
Comment (iFurySt) [reply opendevin/sandbox/sandbox.py:14]: oh, my mistake, update it right now.

File: .github/workflows/ghcr.yml (hunk 4 above)
Comment (iFurySt) [reply .github/workflows/ghcr.yml:16]: I have no idea about such trigger condition, I'll try to find it is possible to do that later.

Comment (rbren) [reply .github/workflows/ghcr.yml:16]: We'll want to commit every time the _code_ changes. But we definitely want to limit it to the `main` branch, or we'll be overwriting it with our feature branches

File: evaluation/SWE-bench/Makefile [hunk 5]
```diff
@@ -0,0 +1,22 @@
+DOCKER_BUILD_REGISTRY=ghcr.io
+DOCKER_BUILD_ORG=opendevin
+DOCKER_BUILD_REPO=eval-swe-bench
+DOCKER_BUILD_TAG=v0.1
+FULL_IMAGE=$(DOCKER_BUILD_REGISTRY)/$(DOCKER_BUILD_ORG)/$(DOCKER_BUILD_REPO):$(DOCKER_BUILD_TAG)
```
Comment (rbren) [inline evaluation/SWE-bench/Makefile:5]: In the future, it'd be great if we `git tag` this repo, the github action applies that tag to the image.

File: .github/workflows/ghcr.yml (hunk 4 above)
Comment (iFurySt) [reply .github/workflows/ghcr.yml:16]: I find some tricks to realize it:
1. detect all Dockerfile are changed. this method will trigger the workflow when any Dockerfile changed.
```
on:
  push:
    branches: [ main ]
    paths:
      - '**/Dockerfile'   // here.
```
2. compare the recent commits to confirm it is need to build. this method need to checkout multi commits(the depth is 2 when we decide to compare the last two commits) and may skip building when multi commits be merged, sth like this:
```
git diff --name-only ${{ github.event.before }} ${{ github.sha }} | grep "$dockerfile_dir/Dockerfile"
if [ $? -eq 0 ]; then
  need_build=true 
  break
fi
```
btw, we also can combine two methods.

how do you think about this?  or do you have any other solutions?

File: evaluation/SWE-bench/Makefile (hunk 5 above)
Comment (iFurySt) [reply evaluation/SWE-bench/Makefile:5]: if we have standard about this, the tag or branch to control is a good way.

File: .github/workflows/ghcr.yml (hunk 4 above)
Comment (xingyaoww) [reply .github/workflows/ghcr.yml:16]: @rbren do you think we need to re-build images on every main branch change, or just `Dockerfile` change? If you think we should rebuild on every git commits, we probably don't need to do this trick?

**Pull request closed without merging**